src/pykeen/sampling/basic_negative_sampler.py

```diff
@@ -49,13 +49,26 @@ def sample(self, positive_batch: torch.LongTensor) -> torch.LongTensor:
         # Sample random entities as replacement
         negative_entities = torch.randint(high=self.num_entities - 1, size=(num_negs,), device=positive_batch.device)
 
-        # Replace heads – To make sure we don't replace the head by the original value
-        # we shift all values greater or equal than the original value by one up
-        # for that reason we choose the random value from [0, num_entities -1]
-        filter_same_head = (negative_entities[:split_idx] >= positive_batch[:split_idx, 0])
-        negative_batch[:split_idx, 0] = negative_entities[:split_idx] + filter_same_head.long()
-        # Corrupt tails
-        filter_same_tail = (negative_entities[split_idx:] >= positive_batch[split_idx:, 2])
-        negative_batch[split_idx:, 2] = negative_entities[split_idx:] + filter_same_tail.long()
+        # Sample random relations as replacement, if requested
+        if 1 in self._corruption_indices:
+            negative_relations = torch.randint(
+                high=self.num_relations - 1,
+                size=(num_negs,),
+                device=positive_batch.device,
+            )
+
+        for index, start in zip(self.corruption_scheme, range(0, num_negs, split_idx)):
+            stop = min(start + split_idx, num_negs)
+            # Replace {heads, relations, tails} – To make sure we don't replace the {head, relation, tail} by the
+            # original value we shift all values greater or equal than the original value by one up
+            # for that reason we choose the random value from [0, num_{heads, relations, tails} -1]
+            if index == 1:
```

Comment:
From the code, both branches look exactly the same besides variable names

Reply:
I also see the same thing

Reply:
no I don't, the thing that's getting sliced is `negative_relations` in the first one and `negative_entities` in the second one

Reply:
Yes, that's why it is different. Can't combine those two 🤷

Reply:
I think I was more referring to something like
```python
negative_ids = negative_relations if index == 1 else negative_entities
negative_batch[start:stop, index] = negative_ids[start:stop]
```
since it is the lhs which is the same.

However, the difference was larger before the other duplicate code disappeared :sweat_smile:

Reply:
I think the way it is currently is more readable, at least to me (so very subjective).
Objectively, both versions are equally efficient, except that the code above creates an additional variable binding.

Reply:
I agree with @lvermue on this one. Here's a third, much *worse* suggestion:

```python
negative_batch[start:stop, index] = (negative_relations if index == 1 else negative_entities)[start:stop]
```